### PR TITLE
Remove deadlock detection / process reconciliation logic

### LINKED ISSUE #12797: Refactor Strategy Script Launcher and Deadlock handling
## Proposed refactor

Currently, if the training scripts are to be spawned from the strategy itself (as in the strategies/ddp.py):
1. It uses '_SubprocessScriptLauncher(self.cluster_environment, self.num_processes, self.num_nodes)'
2. It handles deadlock detection and handling for the launched processes in the strategies/ddp.py file.
We can refactor this logic to allow reuse in other strategies.

### Motivation

This will enable us to share the script launch and deadlock handling logic with the FSDP Native Strategy #12447 or any other strategies in the future.

### Pitch

We can take one of the following approaches
1. Move into parallel strategy (strategies/parallel.py)
2. Move into standalone libraries that are referenced in the strategy

Open to other ideas that come up in the discussion on the issue

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Deadlock detection / process reconciliation is a feature of the DDP strategy family. It is not documented, but it is turned on by default when ddp strategies are used, unless PL_RECONCILIATE_PROCESS=0 is set.

- It has reliability issues, such as that it relies on the current working directory be shared among all nodes, which is often the case but not always. 
- TorchElastic already has mechanisms to detect failing processes and can prevent a hang, and can restart the job automatically. TorchElastic has been upstreamed to PyTorch recently.
- The feature is implemented and tested only for the DDP strategy. Others who inherit from DDP are not tested. Other strategies that inherit from ParallelStrategy do not have this implemented.
- With process reconciliation turned off, an exception that occurs only on one rank but not others will lead to a hang, but the error message surfaces regardless. 
- The main benefit of the feature is that processes get terminated immediately and don't hang. This could save you some money for example when running in cloud instances. 

Part of #16410 
Closes #12797

### Does your PR introduce any breaking changes? If yes, please list them.

Yes, for everyone using the ddp strategy, and anyone else that enables `PL_RECONCILIATE_PROCESS=1`, the feature will no longer work. However, errors 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @justusschock @awaelchli

### CODE REVIEW

**Comment by carmocca** [comment]
I rebased this

File: src/pytorch_lightning/strategies/ddp.py [hunk 1]
```diff
@@ -145,18 +135,12 @@ def _configure_launcher(self) -> None:
         assert self.cluster_environment is not None
         if not self.cluster_environment.creates_processes_externally:
             self._launcher = _SubprocessScriptLauncher(self.cluster_environment, self.num_processes, self.num_nodes)
-            self._rank_0_will_call_children_scripts = True
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/ddp.py:148]: When this ran, `_should_run_deadlock_detection` would return True which means that the deadlock detection logic ran for most people using the subprocessscriptlauncher, not just those who set the environment.

However, it's not clear how useful this was. I just wanted to point out that this was enabled more often than we thought (if I'm not missing something)

**Review by carmocca — APPROVED** [review]
Pushed a commit with some leftover removals to the fsdp strategy

File: src/pytorch_lightning/strategies/ddp.py (hunk 1 above)
Comment (awaelchli) [reply src/pytorch_lightning/strategies/ddp.py:148]: I didn't notice that. I think regardless of this, the deadlock detection does not influence how the program runs in the normal case and so it won't be a breaking change. I will update the description of the PR.

Comment (carmocca) [reply src/pytorch_lightning/strategies/ddp.py:148]: Yes. To me, it's a question of "how useful was this for DDP users?"

if it wasn't then we can proceed.
otherwise maybe we should just convert this logic into a plugin that strategies opt-in. Instead of through inheritance where they are forced to override or deal with this logic

Comment (awaelchli) [reply src/pytorch_lightning/strategies/ddp.py:148]: Created #16410 as a result from our offline chat.

**Review by justusschock — APPROVED** [review]
